File: DTMS365-AppSource/DynamicsTMS365/Models/BookCarrier.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using DTO = Ngl.FreightMaster.Data.DataTransferObjects;
using LTS = Ngl.FreightMaster.Data.LTS;
using DTran = Ngl.Core.Utility.DataTransformation;
using DynamicsTMS365.Controllers;
using NGL.UTC.Library;
using System.Configuration;

namespace DynamicsTMS365.Models
{
    public class BookCarrier
    {
        public int BookControl { get; set; }
        public string BookCarrFBNumber { get; set; }
        public string BookCarrOrderNumber { get; set; }
        public string BookCarrBLNumber { get; set; }
        public string BookCarrBookContact { get; set; }        
        //public DateTime? BookCarrActLoadComplete_Date { get; set; }
        public string BookCarrDockPUAssigment { get; set; }        
        public string BookCarrDockDelAssignment { get; set; }
        public int BookCarrVarDay { get; set; }
        public int BookCarrVarHrs { get; set; }
        public string BookCarrTrailerNo { get; set; }
        public string BookCarrSealNo { get; set; }
        public string BookCarrDriverNo { get; set; }
        public string BookCarrDriverName { get; set; }
        public string BookCarrRouteNo { get; set; }
        public string BookCarrTripNo { get; set; }
        public string BookWhseAuthorizationNo { get; set; }        
        public int BookAMSPickupApptControl { get; set; }
        public int BookAMSDeliveryApptControl { get; set; }
        private string _BookModUser;
        public string BookModUser
        {
            get { return _BookModUser.Left(100); } //uses extension string method Left
            set { _BookModUser = value.Left(100); }
        }

        //Modified by RHR for v-8.5.4.004 on 12/07/2023 date properties converted to strings
        public string BookCarrBookDate { get; set; }
        public string BookCarrScheduleDate { get; set; }
        public string BookCarrActualDate { get; set; }
        public string BookCarrActLoadCompleteDate { get; set; }
        public string BookCarrPODate { get; set; }
        public string BookCarrApptDate { get; set; }
        public string BookCarrActDate { get; set; }
        public string BookCarrActUnloadCompDate { get; set; }
        public string BookCarrStartLoadingDate { get; set; }
        public string BookCarrFinishLoadingDate { get; set; }
        public string BookCarrStartUnloadingDate { get; set; }
        public string BookCarrFinishUnloadingDate { get; set; }
        public string BookModDate { get; set; }





        // time properties to convert
        public string BookCarrBookTime { get; set; }
        public string BookCarrScheduleTime { get; set; }
        public string BookCarrActualTime { get; set; }
        public string BookCarrActLoadCompleteTime { get; set; }
        public string BookCarrPOTime { get; set; }
        public string BookCarrApptTime { get; set; }
        public string BookCarrActTime { get; set; }
        public string BookCarrActUnloadCompTime { get; set; }
        public string BookCarrStartLoadingTime { get; set; }
        public string BookCarrFinishLoadingTime { get; set; }
        public string BookCarrStartUnloadingTime { get; set; }
        public string BookCarrFinishUnloadingTime { get; set; }

        public static string FormatDateToTime(DateTime? val) { if (val.HasValue) { return val.Value.ToShortTimeString(); } else { return ""; } }
        public static DateTime? FormatTimeToDate(DateTime? dtParent, string val)
        {
            DateTime dtcurrent = DateTime.Now;
            DateTime? dtRet = null;
            if (string.IsNullOrWhiteSpace(val)) { return dtRet; }
            if (dtParent.HasValue) { dtcurrent = dtParent.Value; }
            DateTime dtKendoTime;
            string sTime = "";
            if (DateTime.TryParse(val, out dtKendoTime))
            {
                sTime = dtKendoTime.ToShortTimeString();
            }
            string dtString = dtcurrent.ToShortDateString() + " " + sTime;
            DateTime newdateTime;
            if (DateTime.TryParse(dtString, out newdateTime))
            {
                dtRet = newdateTime;
            }
            return dtRet;
        }

        public static Models.BookCarrier selectModelData(DTO.BookCarrier d)
        {
            string serverTimeZone = ConfigurationManager.AppSettings["ServerTimeZone"];
            UsersController userController = new UsersController(); // Create an instance of UserController
            var userSettings = userController.GetUserSettings();

            var userData = (Models.User)userSettings.Data.GetValue(0);

            Models.BookCarrier modelRecord = new Models.BookCarrier();
            if (d != null)
            {
                List<string> skipObjs = new List<string> {
                    "BookCarrBookDate",
                    "BookCarrScheduleDate",
                    "BookCarrActualDate",
                    "BookCarrActLoadCompleteDate",
                    "BookCarrPODate",
                    "BookCarrApptDate",
                    "BookCarrActDate",
                    "BookCarrActUnloadCompDate",
                    "BookCarrStartLoadingDate",
                    "BookCarrFinishLoadingDate",
                    "BookCarrStartUnloadingDate",
                    "BookCarrFinishUnloadingDate",
                    "BookModDate",
                    "BookCarrActLoadComplete_Date", 
                    "BookCarrBookTime", 
                    "BookCarrScheduleTime",
                    "BookCarrActualTime", 
                    "BookCarrActLoadCompleteTime", 
                    "BookCarrPOTime", 
                    "BookCarrApptTime", 
                    "BookCarrActTime", 
                    "BookCarrActUnloadCompTime", 
                    "BookCarrStartLoadingTime", 
                    "BookCarrFinishLoadingTime", 
                    "BookCarrStartUnloadingTime", 
                    "BookCarrFinishUnloadingTime" };
                string sMsg = "";
                modelRecord = (Models.BookCarrier)DTran.CopyMatchingFields(modelRecord, d, skipObjs, ref sMsg);
                if (modelRecord != null) {
                    //Modified by RHR for v-8.5.4.004 on 12/07/2023 changed logic to use strings for dates due to UTC conversion issues
                    modelRecord.BookCarrBookDate = Utilities.convertDateToDateTimeString(d.BookCarrBookDate);
                    modelRecord.BookCarrScheduleDate = Utilities.convertDateToDateTimeString(d.BookCarrScheduleDate);
                    modelRecord.BookCarrActualDate = Utilities.convertDateToDateTimeString(d.BookCarrActualDate);
                    modelRecord.BookCarrPODate = Utilities.convertDateToDateTimeString(d.BookCarrPODate);
                    modelRecord.BookCarrApptDate = Utilities.convertDateToDateTimeString(d.BookCarrApptDate);
                    modelRecord.BookCarrActDate = Utilities.convertDateToDateTimeString(d.BookCarrActDate);
                    modelRecord.BookCarrActUnloadCompDate = Utilities.convertDateToDateTimeString(d.BookCarrStartLoadingDate);
                    modelRecord.BookCarrStartLoadingDate = Utilities.convertDateToDateTimeString(d.BookCarrStartLoadingDate);
                    modelRecord.BookCarrFinishLoadingDate = Utilities.convertDateToDateTimeString(d.BookCarrFinishLoadingDate);
                    modelRecord.BookCarrStartUnloadingDate = Utilities.convertDateToDateTimeString(d.BookCarrStartUnloadingDate);
                    modelRecord.BookCarrFinishUnloadingDate = Utilities.convertDateToDateTimeString(d.BookCarrFinishUnloadingDate);
                    //modelRecord.BookModDate = Utilities.convertDateToDateTimeString(d.BookModDate); Old Functionality
                    modelRecord.BookModDate = clsApplication.convertDateToDateTimeString(d.BookModDate,userData.UserCultureInfo,serverTimeZone,userData.UserTimeZone); //UTC Mod Date Update
                    modelRecord.BookCarrActLoadCompleteDate = Utilities.convertDateToDateTimeString(d.BookCarrActLoadComplete_Date);
                    //Modified by RHR for v-8.5.4.004 on 12/07/2023 changed logic from ToShortTimeString to ToString("HH:mm") to support new UI Requirements
                    //convert time strings for text box
                    modelRecord.BookCarrBookTime = (d.BookCarrBookTime ?? DateTime.Parse("01/01/2021 00:01")).ToString("HH:mm");
                    modelRecord.BookCarrScheduleTime = (d.BookCarrScheduleTime ?? DateTime.Parse("01/01/2021 00:01")).ToString("HH:mm");
                    modelRecord.BookCarrActualTime = (d.BookCarrActualTime ?? DateTime.Parse("01/01/2021 00:01")).ToString("HH:mm");
                    modelRecord.BookCarrActLoadCompleteTime = (d.BookCarrActLoadCompleteTime ?? DateTime.Parse("01/01/2021 00:01")).ToString("HH:mm");
                    modelRecord.BookCarrPOTime = (d.BookCarrPOTime ?? DateTime.Parse("01/01/2021 00:01")).ToString("HH:mm");
                    modelRecord.BookCarrApptTime = (d.BookCarrApptTime ?? DateTime.Parse("01/01/2021 00:01")).ToString("HH:mm");
                    modelRecord.BookCarrActTime = (d.BookCarrActTime ?? DateTime.Parse("01/01/2021 00:01")).ToString("HH:mm");
                    modelRecord.BookCarrActUnloadCompTime = (d.BookCarrActUnloadCompTime ?? DateTime.Parse("01/01/2021 00:01")).ToString("HH:mm");
                    modelRecord.BookCarrStartLoadingTime = (d.BookCarrStartLoadingTime ?? DateTime.Parse("01/01/2021 00:01")).ToString("HH:mm");
                    modelRecord.BookCarrFinishLoadingTime = (d.BookCarrFinishLoadingTime ?? DateTime.Parse("01/01/2021 00:01")).ToString("HH:mm");
                    modelRecord.BookCarrStartUnloadingTime = (d.BookCarrStartUnloadingTime ?? DateTime.Parse("01/01/2021 00:01")).ToString("HH:mm");
                    modelRecord.BookCarrFinishUnloadingTime = (d.BookCarrFinishUnloadingTime ?? DateTime.Parse("01/01/2021 00:01")).ToString("HH:mm");
                }
            }
            return modelRecord;
        }

        public static DTO.BookCarrier selectDTOData(Models.BookCarrier d)
        {
            string serverTimeZone = ConfigurationManager.AppSettings["ServerTimeZone"];
            UsersController userController = new UsersController(); // Create an instance of UserController
            var userSettings = userController.GetUserSettings();
            var userData = (Models.User)userSettings.Data.GetValue(0);

            DTO.BookCarrier dtoRecord = new DTO.BookCarrier();
            if (d != null)
            {
                List<string> skipObjs = new List<string> {
                    "BookCarrBookDate",
                    "BookCarrScheduleDate",
                    "BookCarrActualDate",
                    "BookCarrActLoadCompleteDate",
                    "BookCarrPODate",
                    "BookCarrApptDate",
                    "BookCarrActDate",
                    "BookCarrActUnloadCompDate",
                    "BookCarrStartLoadingDate",
                    "BookCarrFinishLoadingDate",
                    "BookCarrStartUnloadingDate",
                    "BookCarrFinishUnloadingDate",
                    "BookModDate",
                    "BookCarrActLoadComplete_Date",
                    "BookCarrBookTime",
                    "BookCarrScheduleTime",
                    "BookCarrActualTime",
                    "BookCarrActLoadCompleteTime",
                    "BookCarrPOTime",
                    "BookCarrApptTime",
                    "BookCarrActTime",
                    "BookCarrActUnloadCompTime",
                    "BookCarrStartLoadingTime",
                    "BookCarrFinishLoadingTime",
                    "BookCarrStartUnloadingTime",
                    "BookCarrFinishUnloadingTime"
                };
                string sMsg = "";
                dtoRecord = (DTO.BookCarrier)DTran.CopyMatchingFields(dtoRecord, d, skipObjs, ref sMsg);
                if (dtoRecord != null)
                {
                    //Modified by RHR for v-8.5.4.004 on 12/07/2023 changed logic to use strings for dates due to UTC conversion issues
                   dtoRecord.BookCarrBookDate = Utilities.convertStringToNullDateTime(d.BookCarrBookDate);
                   dtoRecord.BookCarrScheduleDate = Utilities.convertStringToNullDateTime(d.BookCarrScheduleDate);
                   dtoRecord.BookCarrActualDate = Utilities.convertStringToNullDateTime(d.BookCarrActualDate);
                   dtoRecord.BookCarrPODate = Utilities.convertStringToNullDateTime(d.BookCarrPODate);
                   dtoRecord.BookCarrApptDate = Utilities.convertStringToNullDateTime(d.BookCarrApptDate);
                   dtoRecord.BookCarrActDate = Utilities.convertStringToNullDateTime(d.BookCarrActDate);
                   dtoRecord.BookCarrActUnloadCompDate = Utilities.convertStringToNullDateTime(d.BookCarrStartLoadingDate);
                   dtoRecord.BookCarrStartLoadingDate = Utilities.convertStringToNullDateTime(d.BookCarrStartLoadingDate);
                   dtoRecord.BookCarrFinishLoadingDate = Utilities.convertStringToNullDateTime(d.BookCarrFinishLoadingDate);
                   dtoRecord.BookCarrStartUnloadingDate = Utilities.convertStringToNullDateTime(d.BookCarrStartUnloadingDate);
                   dtoRecord.BookCarrFinishUnloadingDate = Utilities.convertStringToNullDateTime(d.BookCarrFinishUnloadingDate);
                    //dtoRecord.BookModDate = Utilities.convertStringToNullDateTime(d.BookModDate);
                    dtoRecord.BookModDate = clsApplication.convertStringToNullDateTime(d.BookModDate, userData.UserCultureInfo, userData.UserTimeZone, serverTimeZone); //UTC Mod Date Update
                    dtoRecord.BookCarrActLoadComplete_Date = Utilities.convertStringToNullDateTime(d.BookCarrActLoadCompleteDate);
                    //Modified by RHR for v-8.5.4.003 on 10/28/2023 added new convertStringTimeToDate parser now support 24 hour time format
                    // includes logic for a default value if the user enters bad data in the text box
                    DateTime dtDefault = DateTime.Today;
                    dtoRecord.BookCarrBookTime = Utilities.convertStringTimeToDate(d.BookCarrBookTime,dtoRecord.BookCarrBookDate);
                    dtoRecord.BookCarrScheduleTime = Utilities.convertStringTimeToDate(d.BookCarrScheduleTime,dtoRecord.BookCarrScheduleDate);
                    dtoRecord.BookCarrActualTime = Utilities.convertStringTimeToDate(d.BookCarrActualTime, dtoRecord.BookCarrActualDate);
                    dtoRecord.BookCarrActLoadCompleteTime = Utilities.convertStringTimeToDate(d.BookCarrActLoadCompleteTime, dtoRecord.BookCarrActLoadComplete_Date);
                    dtoRecord.BookCarrPOTime = Utilities.convertStringTimeToDate(d.BookCarrPOTime, dtoRecord.BookCarrPODate);
                    dtoRecord.BookCarrApptTime = Utilities.convertStringTimeToDate(d.BookCarrApptTime,dtoRecord.BookCarrApptDate);
                    dtoRecord.BookCarrActTime = Utilities.convertStringTimeToDate(d.BookCarrActTime, dtoRecord.BookCarrActDate);
                    dtoRecord.BookCarrActUnloadCompTime = Utilities.convertStringTimeToDate(d.BookCarrActUnloadCompTime, dtoRecord.BookCarrActUnloadCompDate);
                    dtoRecord.BookCarrStartLoadingTime = Utilities.convertStringTimeToDate(d.BookCarrStartLoadingTime, dtoRecord.BookCarrStartLoadingDate);
                    dtoRecord.BookCarrFinishLoadingTime = Utilities.convertStringTimeToDate(d.BookCarrFinishLoadingTime, dtoRecord.BookCarrFinishLoadingDate);
                    dtoRecord.BookCarrStartUnloadingTime = Utilities.convertStringTimeToDate(d.BookCarrStartUnloadingTime, dtoRecord.BookCarrStartUnloadingDate);
                    dtoRecord.BookCarrFinishUnloadingTime = Utilities.convertStringTimeToDate(d.BookCarrFinishUnloadingTime, dtoRecord.BookCarrFinishUnloadingDate);
                }
            }
            return dtoRecord;
        }
    }
}

File: NGL.DataAssemblies/Ngl.FreightMaster.Data/DTO/BookData/BookCarrier.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Runtime.Serialization
Imports Ngl.Core.ChangeTracker

Namespace DataTransferObjects
    <DataContract()> _
    Public Class BookCarrier
        Inherits DTOBaseClass


#Region " Data Members"

        Private _BookControl As Integer = 0
        <DataMember()> _
        Public Property BookControl() As Integer
            Get
                Return _BookControl
            End Get
            Set(ByVal value As Integer)
                _BookControl = value
            End Set
        End Property

        Private _BookModDate As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookModDate() As System.Nullable(Of Date)
            Get
                Return _BookModDate
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookModDate = value
            End Set
        End Property

        Private _BookModUser As String = ""
        <DataMember()> _
        Public Property BookModUser() As String
            Get
                Return Left(_BookModUser, 100)
            End Get
            Set(ByVal value As String)
                _BookModUser = Left(value, 100)
            End Set
        End Property

        Private _BookCarrFBNumber As String = ""
        <DataMember()> _
        Public Property BookCarrFBNumber() As String
            Get
                Return Left(_BookCarrFBNumber, 20)
            End Get
            Set(ByVal value As String)
                _BookCarrFBNumber = Left(value, 20)
            End Set
        End Property

        Private _BookCarrOrderNumber As String = ""
        <DataMember()> _
        Public Property BookCarrOrderNumber() As String
            Get
                Return Left(_BookCarrOrderNumber, 20)
            End Get
            Set(ByVal value As String)
                _BookCarrOrderNumber = Left(value, 20)
            End Set
        End Property

        Private _BookCarrBLNumber As String = ""
        <DataMember()> _
        Public Property BookCarrBLNumber() As String
            Get
                Return Left(_BookCarrBLNumber, 20)
            End Get
            Set(ByVal value As String)
                _BookCarrBLNumber = Left(value, 20)
            End Set
        End Property

        Private _BookCarrBookDate As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrBookDate() As System.Nullable(Of Date)
            Get
                Return _BookCarrBookDate
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrBookDate = value
            End Set
        End Property

        Private _BookCarrBookTime As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrBookTime() As System.Nullable(Of Date)
            Get
                Return _BookCarrBookTime
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrBookTime = value
            End Set
        End Property

        Private _BookCarrBookContact As String = ""
        <DataMember()> _
        Public Property BookCarrBookContact() As String
            Get
                Return Left(_BookCarrBookContact, 50)
            End Get
            Set(ByVal value As String)
                _BookCarrBookContact = Left(value, 50)
            End Set
        End Property

        Private _BookCarrScheduleDate As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrScheduleDate() As System.Nullable(Of Date)
            Get
                Return _BookCarrScheduleDate
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrScheduleDate = value
            End Set
        End Property

        Private _BookCarrScheduleTime As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrScheduleTime() As System.Nullable(Of Date)
            Get
                Return _BookCarrScheduleTime
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrScheduleTime = value
            End Set
        End Property

        Private _BookCarrActualDate As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrActualDate() As System.Nullable(Of Date)
            Get
                Return _BookCarrActualDate
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrActualDate = value
            End Set
        End Property

        Private _BookCarrActualTime As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrActualTime() As System.Nullable(Of Date)
            Get
                Return _BookCarrActualTime
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrActualTime = value
            End Set
        End Property

        Private _BookCarrActLoadComplete_Date As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrActLoadComplete_Date() As System.Nullable(Of Date)
            Get
                Return _BookCarrActLoadComplete_Date
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrActLoadComplete_Date = value
            End Set
        End Property

        Private _BookCarrActLoadCompleteTime As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrActLoadCompleteTime() As System.Nullable(Of Date)
            Get
                Return _BookCarrActLoadCompleteTime
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrActLoadCompleteTime = value
            End Set
        End Property

        Private _BookCarrDockPUAssigment As String = ""
        <DataMember()> _
        Public Property BookCarrDockPUAssigment() As String
            Get
                Return Left(_BookCarrDockPUAssigment, 10)
            End Get
            Set(ByVal value As String)
                _BookCarrDockPUAssigment = Left(value, 10)
            End Set
        End Property

        Private _BookCarrPODate As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrPODate() As System.Nullable(Of Date)
            Get
                Return _BookCarrPODate
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrPODate = value
            End Set
        End Property

        Private _BookCarrPOTime As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrPOTime() As System.Nullable(Of Date)
            Get
                Return _BookCarrPOTime
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrPOTime = value
            End Set
        End Property

        Private _BookCarrApptDate As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrApptDate() As System.Nullable(Of Date)
            Get
                Return _BookCarrApptDate
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrApptDate = value
            End Set
        End Property

        Private _BookCarrApptTime As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrApptTime() As System.Nullable(Of Date)
            Get
                Return _BookCarrApptTime
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrApptTime = value
            End Set
        End Property

        Private _BookCarrActDate As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrActDate() As System.Nullable(Of Date)
            Get
                Return _BookCarrActDate
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrActDate = value
            End Set
        End Property

        Private _BookCarrActTime As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrActTime() As System.Nullable(Of Date)
            Get
                Return _BookCarrActTime
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrActTime = value
            End Set
        End Property

        Private _BookCarrActUnloadCompDate As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrActUnloadCompDate() As System.Nullable(Of Date)
            Get
                Return _BookCarrActUnloadCompDate
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrActUnloadCompDate = value
            End Set
        End Property

        Private _BookCarrActUnloadCompTime As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrActUnloadCompTime() As System.Nullable(Of Date)
            Get
                Return _BookCarrActUnloadCompTime
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrActUnloadCompTime = value
            End Set
        End Property

        Private _BookCarrDockDelAssignment As String = ""
        <DataMember()> _
        Public Property BookCarrDockDelAssignment() As String
            Get
                Return Left(_BookCarrDockDelAssignment, 10)
            End Get
            Set(ByVal value As String)
                _BookCarrDockDelAssignment = Left(value, 10)
            End Set
        End Property

        Private _BookCarrVarDay As Integer = 0
        <DataMember()> _
        Public Property BookCarrVarDay() As Integer
            Get
                Return _BookCarrVarDay
            End Get
            Set(ByVal value As Integer)
                _BookCarrVarDay = value
            End Set
        End Property

        Private _BookCarrVarHrs As Integer = 0
        <DataMember()> _
        Public Property BookCarrVarHrs() As Integer
            Get
                Return _BookCarrVarHrs
            End Get
            Set(ByVal value As Integer)
                _BookCarrVarHrs = value
            End Set
        End Property

        Private _BookCarrTrailerNo As String = ""
        <DataMember()> _
        Public Property BookCarrTrailerNo() As String
            Get
                Return Left(_BookCarrTrailerNo, 50)
            End Get
            Set(ByVal value As String)
                _BookCarrTrailerNo = Left(value, 50)
            End Set
        End Property

        Private _BookCarrSealNo As String = ""
        <DataMember()> _
        Public Property BookCarrSealNo() As String
            Get
                Return Left(_BookCarrSealNo, 50)
            End Get
            Set(ByVal value As String)
                _BookCarrSealNo = Left(value, 50)
            End Set
        End Property

        Private _BookCarrDriverNo As String = ""
        <DataMember()> _
        Public Property BookCarrDriverNo() As String
            Get
                Return Left(_BookCarrDriverNo, 50)
            End Get
            Set(ByVal value As String)
                _BookCarrDriverNo = Left(value, 50)
            End Set
        End Property

        Private _BookCarrDriverName As String = ""
        <DataMember()> _
        Public Property BookCarrDriverName() As String
            Get
                Return Left(_BookCarrDriverName, 50)
            End Get
            Set(ByVal value As String)
                _BookCarrDriverName = Left(value, 50)
            End Set
        End Property

        Private _BookCarrRouteNo As String = ""
        <DataMember()> _
        Public Property BookCarrRouteNo() As String
            Get
                Return Left(_BookCarrRouteNo, 50)
            End Get
            Set(ByVal value As String)
                _BookCarrRouteNo = Left(value, 50)
            End Set
        End Property

        Private _BookCarrTripNo As String = ""
        <DataMember()> _
        Public Property BookCarrTripNo() As String
            Get
                Return Left(_BookCarrTripNo, 50)
            End Get
            Set(ByVal value As String)
                _BookCarrTripNo = Left(value, 50)
            End Set
        End Property

        Private _BookWhseAuthorizationNo As String = ""
        <DataMember()> _
        Public Property BookWhseAuthorizationNo() As String
            Get
                Return Left(_BookWhseAuthorizationNo, 20)
            End Get
            Set(ByVal value As String)
                _BookWhseAuthorizationNo = Left(value, 20)
            End Set
        End Property

        Private _BookCarrStartLoadingDate As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrStartLoadingDate() As System.Nullable(Of Date)
            Get
                Return _BookCarrStartLoadingDate
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrStartLoadingDate = value
            End Set
        End Property

        Private _BookCarrStartLoadingTime As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrStartLoadingTime() As System.Nullable(Of Date)
            Get
                Return _BookCarrStartLoadingTime
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrStartLoadingTime = value
            End Set
        End Property

        Private _BookCarrFinishLoadingDate As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrFinishLoadingDate() As System.Nullable(Of Date)
            Get
                Return _BookCarrFinishLoadingDate
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrFinishLoadingDate = value
            End Set
        End Property

        Private _BookCarrFinishLoadingTime As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrFinishLoadingTime() As System.Nullable(Of Date)
            Get
                Return _BookCarrFinishLoadingTime
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrFinishLoadingTime = value
            End Set
        End Property

        Private _BookCarrStartUnloadingDate As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrStartUnloadingDate() As System.Nullable(Of Date)
            Get
                Return _BookCarrStartUnloadingDate
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrStartUnloadingDate = value
            End Set
        End Property

        Private _BookCarrStartUnloadingTime As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrStartUnloadingTime() As System.Nullable(Of Date)
            Get
                Return _BookCarrStartUnloadingTime
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrStartUnloadingTime = value
            End Set
        End Property

        Private _BookCarrFinishUnloadingDate As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrFinishUnloadingDate() As System.Nullable(Of Date)
            Get
                Return _BookCarrFinishUnloadingDate
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrFinishUnloadingDate = value
            End Set
        End Property

        Private _BookCarrFinishUnloadingTime As System.Nullable(Of Date)
        <DataMember()> _
        Public Property BookCarrFinishUnloadingTime() As System.Nullable(Of Date)
            Get
                Return _BookCarrFinishUnloadingTime
            End Get
            Set(ByVal value As System.Nullable(Of Date))
                _BookCarrFinishUnloadingTime = value
            End Set
        End Property


        Private _BookAMSPickupApptControl As Integer = 0
        <DataMember()> _
        Public Property BookAMSPickupApptControl As Integer
            Get
                Return _BookAMSPickupApptControl
            End Get
            Set(value As Integer)
                _BookAMSPickupApptControl = value
            End Set
        End Property

        Private _BookAMSDeliveryApptControl As Integer = 0
        <DataMember()>
        Public Property BookAMSDeliveryApptControl As Integer
            Get
                Return _BookAMSDeliveryApptControl
            End Get
            Set(value As Integer)
                _BookAMSDeliveryApptControl = value
            End Set
        End Property

        Private _BookFinAPActWgt As Integer = 0
        <DataMember()>
        Public Property BookFinAPActWgt() As Integer
            Get
                Return _BookFinAPActWgt
            End Get
            Set(ByVal value As Integer)
                _BookFinAPActWgt = value
            End Set
        End Property


#End Region

#Region " Public Methods"
        Public Overrides Function Clone() As DTOBaseClass
            Dim instance As New BookCarrier
            instance = DirectCast(MemberwiseClone(), BookCarrier)
            Return instance
        End Function

#End Region

    End Class
End Namespace